Hw02 express

yarqui opened the pull request:
Вітаю. Перевірте, будь ласка.

yarqui committed:
add  joi package

yarqui committed:
add all http methods, HttpError helper

yarqui committed:
add HttpError helper, all http methods

yarqui committed:
fix error handler

yarqui committed:
add decorators

Harryk13 commented on models/contacts.js:
```diff
@@ -1,19 +1,63 @@
+    return null;
+  }
+
+  allContacts[index] = { id: contactId, ...data };
```
Що буде, якщо на зміну пришлють не усі проперті?

yarqui replied:
Спрацює валідація схеми :)
Але я зрозумів, про що ви. Зробив спред попереднього значення, і поверх переписав data.

Harryk13 commented on routes/api/contacts.js:
```diff
@@ -1,25 +1,19 @@
+router.delete("/:contactId", ctrl.removeContact);
 
-module.exports = router
+router.put("/:contactId", validateBody(schemas.addSchema), ctrl.updateContact);
```
як мені змінити тільки одне проперті?

yarqui replied:
Хіба для цього не треба використовувати PATCH? Просто в ДЗ цього не було, тому PATCH не робив.
Можу просто прибрати валідацію схеми із PUT, і тоді буде працювати все і повністю, і частково.

Harryk13 replied:
Цитата з ДЗ:

1.     Отримує параметр id
2.     Отримує body в json-форматі c оновленням будь-яких полів name, email и phone
3.     Якщо body немає, повертає json з ключем {"message": "missing fields"} і статусом 400
4.     Якщо з body всі добре, викликає функцію updateContact(contactId, body). (Напиши її) для поновлення контакту в файлі contacts.json
5.     За результатом роботи функції повертає оновлений об'єкт контакту і статусом 200. В іншому випадку, повертає json з ключем "message": "Not found" і статусом 404

Harryk13 replied:
Тобто потрбна можливість зміни будь якого проперті, але при цьому треба і валідація

yarqui replied:
Дякую. Оновлено.

yarqui committed:
🔧tune PUT method

- remove schema validation from PUT method
- fix PUT method to be able to work with partial change of properties

Harryk13 commented on routes/api/contacts.js:
```diff
@@ -14,6 +14,6 @@ router.post("/", validateBody(schemas.addSchema), ctrl.addContact);
 router.delete("/:contactId", ctrl.removeContact);
 
-router.put("/:contactId", validateBody(schemas.addSchema), ctrl.updateContact);
+router.put("/:contactId", ctrl.updateContact);
```
тобто я можу навіть ід змінти? Або будь яке проперті додати?

yarqui committed:
➕ add updateScheme validation

yarqui committed:
➕ add regex for phone validation

yarqui committed:
activate regex
